src/job-queue.cpp

```diff
@@ -0,0 +1,51 @@
+#include <job-queue.h>
+#include <thread>
+
+JobQueue& JobQueue::GetInstance() {
+    static JobQueue jq;
+    return jq;
+}
+
+bool JobQueue::AddJob(std::function<void()> job) {
+    // check the atomic bool, if it indicates the job queue is still running we can add to the queue
+    if(keep_running.load()){
+        queue_lock.lock();
+        jqueue.emplace(job);
+        queue_lock.unlock();
+        cv.notify_all();
+        has_work = true;
+        return true;
+    }
+    // we return false to indicate the queue didn't accept the job
+    return false;
+}
+
+void JobQueue::Process() {
+    keep_running = true;
+    std::mutex wait_mtx;
+    std::unique_lock<std::mutex> wait_lock(wait_mtx);
+    // so long as this stays true we're gonna keep looping
+    while(keep_running.load()){
+        if(!has_work.load()){
```

Comment:
Why don't we just do `if(!jqueue.empty())` ?

Reply:
Cause we have to lock a mutex before we access the jqueue.
As it stands it will wait until there is work to do because it is notified when something is added to the queue and we don't need to look at jqueue directly because we sort of cache the state in the atomic boolean has_work.

To go the other way, we would need some other way to wait for work or we'd have an ultra hot loop running constantly doing nothing but locking a mutex, checking the state, and then unlocking the mutex. The code for that wouldn't be as elegant.

Reply:
Take a look here https://github.com/BCCF-UBCO-AD/Orthanc-TMI/blob/0bc335d1d82b0d49861329a7647216d66b5c7b1b/src/job-queue.cpp to see what it looked like before the conditional variable (which sends the notification of work) in this I sleep the thread, and since I don't know when there will be work I set it to 1s so it does what I described above every second.

Reply:
Gotcha, make sense

Reply:
You know what though, it should be a while loop though and not an if.